Libraries/Vibration/RCTVibration.m

```diff
@@ -20,4 +27,77 @@ @implementation RCTVibration
   AudioServicesPlaySystemSound(kSystemSoundID_Vibrate);
 }
 
+RCT_EXPORT_METHOD(vibrateByPattern:(NSArray *)pattern
+                  repeat:(BOOL)repeat
+                  initialVibrate:(BOOL)initialVibrate)
+{
+  if (_vibrating) {
+    return;
+  }
+  _vibrating = YES;
+  _pattern = pattern;
+  _repeat = repeat;
+  _index = 0;
+  if (initialVibrate) {
+    [self vibrate];
+  }
+  [self vibrateDelayedWithInterval:
+        [RCTConvert NSTimeInterval:_pattern[_index]] ?: 1.0];
+}
+
+RCT_EXPORT_METHOD(cancel)
+{
+  _vibrating = NO;
+  [self deleteTimer];
+}
+
+- (void)vibrateDelayedWithInterval:(NSTimeInterval)interval
+{
+  if (_timer) {
+    [self deleteTimer];
+  }
+  _timer =
+    [NSTimer timerWithTimeInterval:interval
+                                     target:self
+                                   selector:@selector(_onTick:)
+                                   userInfo:nil
+                                    repeats:NO];
+  [[NSRunLoop mainRunLoop] addTimer:_timer
+                            forMode:NSRunLoopCommonModes];
+}
+
+- (void)_onTick:(NSTimer *)timer
+{
+  if (!_vibrating) {
+    return;
+  }
```

Comment:
You'll need something like _RCTTimingProxy (https://github.com/facebook/react-native/blob/master/React/Modules/RCTTiming.m#L73)